hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/PrecompileArgumentsDecoder.java

```diff
@@ -0,0 +1,334 @@
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.TokenID;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.datatypes.Address;
+import org.web3j.abi.FunctionReturnDecoder;
+import org.web3j.abi.TypeReference;
+import org.web3j.abi.datatypes.StaticArray;
+import org.web3j.abi.datatypes.Type;
+import org.web3j.abi.datatypes.generated.Bytes32;
+import org.web3j.abi.datatypes.generated.Int64;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public final class PrecompileArgumentsDecoder {
+
+	private static final int INPUT_ENTRY_BYTES_LENGTH = 32;
+	private static final int ADDRESS_BYTES_LENGTH = 20;
+	private static final int ADDRESS_SKIP_BYTES_LENGTH = 12;
+	private static final int FUNCTION_SELECTOR_BYTES_LENGTH = 4;
+	private static final int MAX_STATIC_ARRAY_LENGTH = 32;
+
+	private PrecompileArgumentsDecoder() {
+	}
+
+	public static Map<Integer, Object> decodeArgumentsForTransferTokens(final Bytes input) {
+
+		// PARSING TOKEN ADDRESS
+		final var tokenAddress = Address.wrap(input.slice(ADDRESS_SKIP_BYTES_LENGTH + FUNCTION_SELECTOR_BYTES_LENGTH, ADDRESS_BYTES_LENGTH));
+		final var tokenID = EntityIdUtils.tokenParsedFromSolidityAddress(tokenAddress.toArray());
+
+		// PARSING ACCOUNT ADDRESS ARRAY
+		final var addressArrayStartPositionStartIndex =
+				ADDRESS_SKIP_BYTES_LENGTH + FUNCTION_SELECTOR_BYTES_LENGTH + ADDRESS_BYTES_LENGTH;
+		final var addressArrayStartPositionBytes = input.slice(addressArrayStartPositionStartIndex, INPUT_ENTRY_BYTES_LENGTH);
+		final var addressArrayStartPosition =
+				addressArrayStartPositionBytes.toBigInteger().intValue() + FUNCTION_SELECTOR_BYTES_LENGTH;
+		final var addressArraySize = input.slice(addressArrayStartPosition, INPUT_ENTRY_BYTES_LENGTH).toBigInteger();
+		final var addressArraySizeInt = addressArraySize.intValue();
+		final var addressArrayByteLength = addressArraySizeInt * INPUT_ENTRY_BYTES_LENGTH;
+		final var addressArrayByteLengthWithHeader = addressArrayByteLength + INPUT_ENTRY_BYTES_LENGTH;
+
+		final var addresses = decodeAddressArray(addressArraySizeInt,
+				addressArrayStartPosition + INPUT_ENTRY_BYTES_LENGTH,
+				input);
+		final var accountIDs = addresses.stream().map(a -> convertAddressToAccountID(a)).collect(Collectors.toList());
+
+		// PARSING AMOUNT ARRAY
+		final var amountArrayStartPosition = addressArrayStartPositionBytes.toBigInteger().intValue() + FUNCTION_SELECTOR_BYTES_LENGTH + addressArrayByteLengthWithHeader;
+		final var amountArraySize = input.slice(amountArrayStartPosition, INPUT_ENTRY_BYTES_LENGTH).toBigInteger();
+		final var amountArraySizeInt = amountArraySize.intValue();
+
+		final var amounts = decodeAmountsArray(amountArraySizeInt, amountArrayStartPosition, input);
+
+		final Map<Integer, Object> decodedArguments = new HashMap<>();
+		decodedArguments.put(1, tokenID);
```

Comment:
Imo it will be a lot more readable and maintainable if we return a `class` for each of these `decodeArgumentsFor*()` methods; e.g. here we could return an instance [of `SyntheticTxnFactory.FungibleTokenTransfer`](FungibleTokenTransfer).